lighthouse-core/gather/gatherers/dobetterweb/optimized-images.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ /**
+  * @fileoverview Determines optimized jpeg/webp filesizes for all same-origin and dataURI images.
+  */
+'use strict';
+
+const Gatherer = require('../gatherer');
+const URL = require('../../../lib/url-shim');
+
+/* global document */
+
+/* istanbul ignore next */
+function getOptimizedNumBytes(url) {
+  return new Promise(function(resolve, reject) {
+    const img = new Image();
+    const canvas = document.createElement('canvas');
+    const context = canvas.getContext('2d');
+
+    function getTypeStats(type, quality) {
+      const dataURI = canvas.toDataURL(type, quality);
+      const base64 = dataURI.slice(dataURI.indexOf(',') + 1);
+      return {base64: base64.length, binary: atob(base64).length};
```

Comment:
Haha no worries :) Yeah I'm actually purposefully calculating the data URI size here (though I could check first to see if it wasn't a data URI to begin with and skip it) so that I can compare it under the assumption if you're shipping an image as a data URI you'd likely send the optimized version as one as well. I suppose avoiding large data URI's should really be an audit  by itself too though 😄